src/audit/cache_poisoning.rs

```diff
@@ -0,0 +1,292 @@
+                // Not sure about what to do here!
+                None
+            }
+            Some(EnvValue::String(value)) => match ExplicitExpr::from_curly(value) {
+                None => {
+                    if action.cache_control_input_toggle == CacheControlToggle::MultiValue {
+                        Some(CacheUsage::DirectOptIn)
+                    } else {
+                        None
+                    }
+                }
+                Some(_) => Some(CacheUsage::ConditionalOptIn),
+            },
+        }
+    }
+
+    fn evaluate_cache_usage(&self, uses: &str, env: &Env) -> Option<CacheUsage> {
+        let (reference, _) = uses.split_once("@")?;
+
+        let known_action = KNOWN_CACHE_AWARE_ACTIONS
+            .iter()
+            .find(|action| action.reference == reference)?;
+
+        let cache_control_input = env
+            .keys()
+            .find(|k| *k == known_action.cache_control_input_name);
+
+        match cache_control_input {
+            // when not using the specific Action input to control caching behaviour,
+            // we evaluate whether it uses caching by default
+            None => CachePoisoning::evaluate_default_action_behaviour(known_action),
+
+            // otherwise, we infer from the value assigned to the cache control input
+            Some(key) => {
+                // first, we extract the value assigned to that input
+                let declared_usage =
+                    CachePoisoning::evaluate_user_defined_opt_in(key, env, known_action);
+
+                // we now evaluate the extracted value against the opt-in semantics
+                match &declared_usage {
+                    Some(CacheUsage::DirectOptIn) => {
+                        if known_action.cache_control_input_defines_opt_in {
+                            declared_usage
+                        } else {
+                            // in this case, the user opted for disabling the cache
+                            // hence we don't return a CacheUsage
+                            None
+                        }
+                    }
+                    // CachePoisoning::evaluate_user_defined_opt_in returns only
+                    // CacheUsage::DirectOptIn and CacheUsage::ConditionalOptIn variants
+                    //
+                    // Because we can't evaluate expressions, there is nothing to do
+                    // regarding CacheUsage::ConditionalOptIn
+                    _ => declared_usage,
+                }
+            }
+        }
+    }
+}
+
+impl WorkflowAudit for CachePoisoning {
+    fn new(_: AuditState) -> anyhow::Result<Self>
+    where
+        Self: Sized,
+    {
+        Ok(Self)
+    }
+
+    fn audit_step<'w>(&self, step: &Step<'w>) -> anyhow::Result<Vec<Finding<'w>>> {
+        let mut findings = vec![];
+
+        let trigger = &step.workflow().on;
+
+        if !self.trigger_used_when_publishing_artifacts(trigger) {
```

Comment:
This looks good for now, thanks!

Thinking out loud: another future proxy we could use for "is this a publishing workflow" is whether or not the workflow contains any well-known publishing actions, e.g. `gh-action-pypi-publish`. 

Not something to do with this iteration, but worth thinking about 🙂